worker/updateChorals.js

```diff
@@ -1,43 +1,71 @@
 var client = require('redis').createClient(process.env.REDIS_STORE_URI)
 var Choral = require('../models/choral')
 var request = require('request')
+var sandbox = new require('sandbox')
 
 setInterval(() => {
   // this doesnt run the chorals according to their timeouts it just runs 
   // all of the chorals every UPDATE_NEW_CHORALS seconds
-  Choral.getAllChoralsWithChildren((err, chorals) => {
+  Choral.getAllChorals((err, chorals) => {
     if(err) return console.log(err.message)
 
     chorals.forEach((choral) => {
+      if(choral.children.length == 0){
+        runComputation({}, choral)
+      }
+
       var param = {}
 
       choral.children.forEach((child) => {
         client.hgetall(child.choralId, (err, data) => {
           if(err) return console.log(err.message)
 
           param[child.name] = data;
-          console.log(data)
-
           // this is just so I dont have to import a promise library
           if(Object.keys(param).length == choral.children.length){
-            // In a real project this would kick off in a sandbox. If we find
-            // the server is crashing a bunch then we can look into doing that
             // Amazon lambda functions would make this a viable solution to
             // our problem because they do not have access to our environment
-            try{
-              // horiffic use of eval
-              eval("var func = " + choral.func)
-
-              func(param, get_callBack(choral))
-            } catch (e) {
-              sendChoralData(choral, { error: e.message })
-            }
+            // if this ends up being super slow we should do that
+            runComputation(param, choral)
           }
         })
       })
     })
   })
-}, 0 | process.env.UPDATE_NEW_CHORALS)
+}, (0 | process.env.UPDATE_NEW_CHORALS) || 5 )
+
+function runComputation(children, choral){
+  const {VM} = require('vm2');
+
+  var config = {
+    func: choral.func,
+    children: children,
+    result: { },
+  }
+
+  const vm = new VM({
+    require: {
+      external: true
+    },
+    sandbox: config,
+    // this gives us the option to use a timeout if we see fit
+    timeout: (0 | process.env.CHORAL_FUNCTION_TIMEOUT) || 50
+  });
+
+  // this callback must be inside of the vm
+  // Note: this is synchronous meaning that the call will block until
+  // this is call is finished.
+  try{
+    vm.run(`
+      const callback = function(res) {
+        result.output = res;
```

Comment:
This is the stupidest and hackiest part of this whole thing. You cant set result you have to set a member of result otherwise it doesnt get output.